File: clients/java/client-impl/src/main/java/com/flipkart/grayskull/hooks/DefaultRefreshHandlerRef.java
package com.flipkart.grayskull.hooks;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.concurrent.atomic.AtomicBoolean;

/**
 * {@link RefreshHandlerRef} returned by {@code registerRefreshHook}.
 */
public final class DefaultRefreshHandlerRef implements RefreshHandlerRef {

    private static final Logger log = LoggerFactory.getLogger(DefaultRefreshHandlerRef.class);

    private final String secretRef;
    private final Runnable onUnregister;
    private final AtomicBoolean active = new AtomicBoolean(true);

    public DefaultRefreshHandlerRef(String secretRef, Runnable onUnregister) {
        this.secretRef = secretRef;
        this.onUnregister = onUnregister;
    }

    @Override
    public String getSecretRef() {
        return secretRef;
    }

    @Override
    public boolean isActive() {
        return active.get();
    }

    @Override
    public void unRegister() {
        if (!active.compareAndSet(true, false)) {
            return;
        }
        onUnregister.run();
        log.debug("Unregistered refresh hook for secretRef:{}", secretRef);
    }
}

File: clients/go/client-impl/internal/hooks/defaultRefreshHandlerRef.go
package hooks

import (
	"log/slog"
	"sync/atomic"

	apiHooks "github.com/flipkart-incubator/grayskull/clients/go/client-api/hooks"
)

// DefaultRefreshHandlerRef is  returned by RegisterRefreshHook
// when the background poller is wired up. Unregister is idempotent
type DefaultRefreshHandlerRef struct {
	secretRef    string
	active       atomic.Bool
	onUnregister func()
}

var _ apiHooks.RefreshHandlerRef = (*DefaultRefreshHandlerRef)(nil)

// NewDefaultRefreshHandlerRef returns a handle bound to secretRef. onUnregister
// runs at most once, on the first Unregister call.
func NewDefaultRefreshHandlerRef(secretRef string, onUnregister func()) *DefaultRefreshHandlerRef {
	r := &DefaultRefreshHandlerRef{secretRef: secretRef, onUnregister: onUnregister}
	r.active.Store(true)
	return r
}

func (r *DefaultRefreshHandlerRef) GetSecretRef() string { return r.secretRef }
func (r *DefaultRefreshHandlerRef) IsActive() bool       { return r.active.Load() }

func (r *DefaultRefreshHandlerRef) Unregister() {
	if !r.active.CompareAndSwap(true, false) {
		return
	}
	if r.onUnregister != nil {
		r.onUnregister()
	}
	slog.Debug("Unregistered refresh hook", "secretRef", r.secretRef)
}
